consensus/src/quorum_store/batch_aggregator.rs

```diff
@@ -0,0 +1,227 @@
+    max_batch_bytes: usize,
+}
+
+impl BatchAggregator {
+    pub(crate) fn new(max_batch_bytes: usize) -> Self {
+        Self {
+            batch_id: None,
+            batch_state: None,
+            max_batch_bytes,
+        }
+    }
+
+    fn next_fragment_id(&self) -> usize {
+        match &self.batch_state {
+            Some(state) => state.num_fragments(),
+            None => 0,
+        }
+    }
+
+    fn outdated_fragment(&self, batch_id: BatchId, fragment_id: usize) -> bool {
+        if let Some(self_batch_id) = self.batch_id {
+            let next_fragment_id = self.next_fragment_id();
+            if next_fragment_id == 0 {
+                // We always append transactions when we create a new state.
+                debug_assert!(
+                    self.batch_state.is_none(),
+                    "No fragments in batch aggregator state"
+                );
+
+                // In this case, the next fragment must start self_batch_id + 1
+                batch_id <= self_batch_id
+            } else {
+                (batch_id, fragment_id) < (self_batch_id, self.next_fragment_id())
+            }
+        } else {
+            false
+        }
+    }
+
+    // Should only be called with a non-outdated fragment.
+    fn missed_fragment(&self, batch_id: BatchId, fragment_id: usize) -> bool {
+        match self.batch_id {
+            Some(self_batch_id) => {
+                if batch_id > self_batch_id {
+                    self.batch_state.is_some() || fragment_id > 0
+                } else {
+                    assert!(
+                        batch_id == self_batch_id,
+                        "Missed fragment called with an outdated fragment"
+                    );
+                    fragment_id > self.next_fragment_id()
+                }
+            }
+            // Allow larger batch_id (> 0) the first time as quorum store might
+            // be recovering from a crash and continuing with a larger batch_id.
+            None => fragment_id > 0,
+        }
+    }
+
+    /// Appends transactions from a batch fragment, ensuring that the fragment is
+    /// consistent with the state being aggregated, and the maximum batch size is
+    /// being respected. Otherwise, a corresponding error is returned which should
+    /// be handled on the caller side (i.e. panic for self, log & ignore for peers).
+    pub(crate) fn append_transactions(
+        &mut self,
+        batch_id: BatchId,
+        fragment_id: usize,
+        transactions: Vec<SerializedTransaction>,
+    ) -> Result<(), BatchAggregatorError> {
+        if self.outdated_fragment(batch_id, fragment_id) {
+            counters::EXPIRED_BATCH_FRAGMENTS_COUNT.inc();
+            // Replay or batch / fragment received out of order.
+            return Err(BatchAggregatorError::OutdatedFragment);
+        }
+
+        let missed_fragment = self.missed_fragment(batch_id, fragment_id);
+        if missed_fragment {
```

Comment:
if we missed a fragment, but new fragment is not the first one in the batch, why do we clean the batch_state? isn't it better to just ignore the new message, as that is what we anyways do in the end?

Reply:
cool, added that optimization.

One question - currently we return MissedFragment error any time we realize something is missing, including the productive case when we start the higher batch successfully (fragment_id = 0). Then, we also have to have a special case in end_batch for single-fragment batches to succeed in this case.
Would it be more appropriate to only return MissedFragment when current fragment can't be aggregated (i.e. not including higher batch fragment 0) - for sure with a proper comment and/or maybe renaming the error kind. This would get rid of missed_fragment bool in append_transactions alltogether, and that special case in end_batch.

Let me know what you think?

Reply:
Actually I went ahead and did this change.